### ISSUE TITLE
What is the value?

### ISSUE THREAD

**Issue by taras** [cowboyd/microstates.js#15]
This issue is the continuation of the conversation from #e-microstates channel.

Here is the transcript of the conversation:

**taras**: One thing that I'm not clear about is what happens with the value when it's boxed in. Is a microstate permanently converting a value it receives or is it type casting it internally and exporting the original value?

**cowboyd**: it “captures” the value, but it does massage it somehow, but only when necessary

 ```js
let object = {};
let state = new State(object);
state.valueOf() === object //=> true
```

However, this will not be the case if a merge happens, but maybe it should.

 ```js
const Thing = State.extend({
  one: 'two'
});
let object = {};
let thing = new Thing(object);
thing.valueOf() //=> {one: 'two'};
thing.valueOf() === object // false
```

But now that I think of it, maybe value of should be `{}` because it does not represent any change

The other case is what do you do when you have a nested state:

```js
let object = {};
const Thing = State.extend({
  one: new State(1),
});

let thing = new Thing(object);
thing.valueOf() //=> {one: 1}
```

but again, maybe that’s not right. Maybe `thing.valueOf()` should be `object`, because there has been no change.

It might be worth having a `materialize` function that collects the values of all properties on the whole tree.

**Comment by taras** [comment]
> But now that I think of it, maybe value of should be {} because it does not represent any change

This will be very confusing. We want composition to work like classes. 

```js
class A {
  foo = 'bar'
}

let a = new A();
a.foo === 'bar' // true
```

The above is expected, and how Microstates work, we should keep this symmetry.

> It might be worth having a `materialize` function that collects the values of all properties on the whole tree.

Is `materialize` a concept that is used anywhere other than the old version of Ember Data?

What about `serialize` and `deserialize`?

* `serialize` will collect all of the values and return a hash. 
* `deserialize` would take a value and return the appropriate value for that microstate value.

We would use this method to type cast Bool

```
const Bool = State.extend({
  deserialize(value) {
    return !!value;
  }
});
```

Deserializing composed microstates would need to happen outside of `deserialize`. 

What do you think about this?

**Comment by taras** [comment]
`serialize` hook would receive original value (which would be lost if microstate is transitioned) and return that value.

```js
let object = {
  "visibility": "0"
};
const Thing = State.extend({
  visibility: new Bool(true)
});

let thing = new Thing(object);
thing.visibility.valueOf() // => false
thing.valueOf() //=> {"visibility": "0"}
```

**Comment by cowboyd** [comment]
Hmm.. this is something to consider. Although the current behavior is:

```js
let object = {
  "visibility": "0"
};
const Thing = State.extend({
  visibility: new Bool(true)
});

let thing = new Thing(object);
thing.valueOf() // => { visibility: false }
```

Which is lossy, but more in line with what I would expect. What about this case?

```js
const Thing = State.extend({
  visibility: new State({
    isVisible: new Bool(false),
    reason: new Str("hidden for quality purposes")
  })
})
new Thing({visibility: {isVisible: "1"}}).valueOf() //=> ???;
```

**Comment by taras** [comment]
```js
let thing = new Thing({visibility: {isVisible: "1"}})
thing.valueOf() //=> {visibility: {isVisible: "1", reason: "hidden for quality purposes"}}

thing.visibility.isVisible.toggle();
thing.valueOf() //=> { visibility: {isVisible: false, reason: "hidden for quality purposes" }}
```

**Comment by cowboyd** [comment]
This has been resolved with #14
